File: blockchain/structures/block.py
import json
import hashlib
import time
from dataclasses import dataclass
from typing import Dict, List

from proofs import AggregatedLeaderRotationProof, AggregatedEpochFinalizationProof
from transaction import Transaction


def get_utc_timestamp_millis():
    return int(time.time() * 1000)


def generate_signature(private_key: str, message: str) -> str:
    # stub
    return f"sig({message[:10]}...)"

def verify_signature(message: str, pubkey: str, signature: str) -> bool:
    # stub
    return signature.startswith("sig(")


@dataclass
class DelayedTxsBatch:
    epochIndex: int
    delayedTransactions: List[Dict[str, str]]
    proofs: Dict[str, str]


@dataclass
class ExtraData:
    rest: Dict[str, str]
    aefpForPreviousEpoch: AggregatedEpochFinalizationProof
    delayedTxsBatch: DelayedTxsBatch
    aggregatedLeadersRotationProofs: Dict[str, AggregatedLeaderRotationProof] # leader -> proof


@dataclass
class Block:
    creator: str
    time: int
    epoch: str
    transactions: List[Transaction]
    extra_data: ExtraData
    index: int
    prev_hash: str
    sig: str = ""

    def get_hash(self,network_id) -> str:
        jsoned_transactions = json.dumps([t.__dict__ for t in self.transactions], sort_keys=True)
        data_to_hash = (
            self.creator +
            str(self.time) +
            jsoned_transactions +
            network_id +
            self.epoch +
            str(self.index) +
            self.prev_hash
        )
        return hashlib.sha256(data_to_hash.encode()).hexdigest()

    def sign_block(self,private_key):
        self.sig = generate_signature(private_key, self.get_hash())

    def verify_signature(self) -> bool:
        return verify_signature(self.get_hash(), self.creator, self.sig)

File: blockchain_golang/block/block.go
package block

import (
	"encoding/json"
	"strconv"

	"github.com/KlyntarNetwork/Web1337Golang/crypto_primitives/ed25519"
	"github.com/VladChernenko/UndchainCore/globals"
	"github.com/VladChernenko/UndchainCore/structures"
	"github.com/VladChernenko/UndchainCore/utils"
)

type ExtraData struct {
	Rest                            map[string]string                                    `json:"rest"`
	AefpForPreviousEpoch            *structures.AggregatedEpochFinalizationProof         `json:"aefpForPreviousEpoch"`
	DelayedTransactionsBatch        structures.DelayedTransactionsBatch                  `json:"delayedTxsBatch"`
	AggregatedLeadersRotationProofs map[string]*structures.AggregatedLeaderRotationProof `json:"aggregatedLeadersRotationProofs"`
}

type Block struct {
	Creator      string                   `json:"creator"`
	Time         int64                    `json:"time"`
	Epoch        string                   `json:"epoch"`
	Transactions []structures.Transaction `json:"transactions"`
	ExtraData    ExtraData                `json:"extraData"`
	Index        int                      `json:"index"`
	PrevHash     string                   `json:"prevHash"`
	Sig          string                   `json:"sig"`
}

func NewBlock(transactions []structures.Transaction, extraData ExtraData, epochFullID string) *Block {
	return &Block{
		Creator:      globals.CONFIGURATION.PublicKey,
		Time:         utils.GetUTCTimestampInMilliSeconds(),
		Epoch:        epochFullID,
		Transactions: transactions,
		ExtraData:    extraData,
		Index:        globals.GENERATION_THREAD_METADATA_HANDLER.NextIndex,
		PrevHash:     globals.GENERATION_THREAD_METADATA_HANDLER.PrevHash,
		Sig:          "",
	}
}

func (block *Block) GetHash() string {

	jsonedTransactions, _ := json.Marshal(block.Transactions)

	networkID := globals.GENESIS.NetworkId

	dataToHash := block.Creator + strconv.FormatInt(block.Time, 10) + string(jsonedTransactions) + networkID + block.Epoch + strconv.FormatUint(uint64(block.Index), 10) + block.PrevHash

	return utils.Blake3(dataToHash)

}

func (block *Block) SignBlock() {

	block.Sig = ed25519.GenerateSignature(globals.CONFIGURATION.PrivateKey, block.GetHash())

}

func (block *Block) VerifySignature() bool {

	return ed25519.VerifySignature(block.GetHash(), block.Creator, block.Sig)

}
